### ISSUE TITLE
torch.zeros(dtype=uint8) is miscompiled

### ISSUE THREAD

**Issue by mgehre-amd** [llvm/torch-mlir#1936]
When doing an e2e test like
```
class ZerosModuleUInt8(torch.nn.Module):

    def __init__(self):
        super().__init__()

    @export
    @annotate_args([
        None,
    ])
    def forward(self):
        return torch.zeros(3, 4, dtype=torch.uint8)


@register_test_case(module_factory=lambda: ZerosModuleUInt8())
def ZerosModuleUInt8_basic(module, tu: TestUtils):
    module.forward()
```
it fails with
```
****** Failed tests - 1 tests
    FAIL - "ZerosModuleUInt8_basic"
        @ trace item #0 - call to "forward"
        @ output of call to "forward"
        ERROR: dtype (torch.int8) is not equal to golden dtype (torch.uint8)
```

I think it's not specific to the `zeros` op; returning `uint8` seems to be the problem. Taking `uint8` as argument seems to work,
and changing the test case to return `int8` also makes it pass.

It think the compiler should either error out on `uint8` or properly support it (preferable).

**Comment by ramiro050** [comment]
Yeah, this is a known issue (see discussion: https://github.com/llvm/torch-mlir/pull/1384#issuecomment-1260011334). Having a check that errors out when the return type of the function is `uint8` would definitely be welcomed. Having full support for `uint8` might be a bit involved, but if it's needed I'm happy to help brainstorm on how to approach it.

**Comment by silvasean** [comment]
I don't think we should have merged https://github.com/llvm/torch-mlir/pull/1384 -- can we revert it? This is a pretty fundamental issue and we shouldn't be merging code that simply ignores the fundamental issue.

**Comment by ramiro050** [comment]
> I don't think we should have merged #1384 -- can we revert it? This is a pretty fundamental issue and we shouldn't be merging code that simply ignores the fundamental issue.

Is a full revert needed? My main worry regarding a full revert is the impact this would have on users of torch-mlir. The change was merged almost 6 months ago, and I think it is fair for users to expect e2e tested features to not disappear after many months.

Moreover, most of the code in the commit seems necessary even if the fundamental issue is fixed. In particular, the code generates correct IR at the backend contract level when the inputs or outputs are `int8/uint8`s. The issue is that the backends and refbackend do not have a way to properly handle `uint8`s. Therefore, I would prefer to simply limit the set of allowed programs at the backend contract to those that don't return `uint8`s. Such programs are currently not tested e2e, so this would not result in the removal of e2e tested features.

In the future we should be more careful about stopping PRs like those from getting merged, or revert early when they do.
